### ISSUE TITLE
Is it possible to use an area around the notch on new 14"/16" MacBooks?

### ISSUE THREAD

**Issue by karpovpw** [MaddTheSane/Simple-Comic#78]
Basically, the title. I really want those sweet vertical pixels.  :(

**Comment by nickv2002** [comment]
I don't have a notchBook yet but maybe will get the new M2 Air and test.

For anyone who does have a notchBook: is the default behavior not ideal?

Background/Ref links:

- https://developer.apple.com/design/human-interface-guidelines/patterns/going-full-screen/
- https://developer.apple.com/documentation/bundleresources/information_property_list/nsprefersdisplaysafeareacompatibilitymode?language=objc
- https://www.macworld.com/article/546104/how-to-macbook-pro-macos-monterey-block-notch-menu-bar.html

**Comment by karpovpw** [comment]
I have a 14".

The default behavior is not using areas around notch (left and right) at all in both compatibility modes. It would be great to see an image stretched all the way up top. Yeah, the notch will be in the way, but we'll get more vertical height and pages will fill entire screen, which is more suitable for manga/comics reading.

My iOS skills are pretty weak, but maybe I can help in any way? I haven't found any useful info online.

**Comment by nickv2002** [comment]
Ack, thanks for info.
One catch about stretching the image above the bottom of the notch is that there's have to be a way to disable it as I can imagine many people would not like that setup.

**Comment by karpovpw** [comment]
Yeah, but simple on/off switch should be enough. Maybe even add a check if laptop screen has more height (ie not 16:10) and show that switch only if a user have notched device.

**Comment by johannkoenig** [comment]
I think a good idea would be to extend the "fit to screen" button so that clicking it again would toggle fitting with the notch. Obviously only use that option when the screen has a notch.

**Comment by blmatthews** [comment]
Are there other apps that do something like that? I haven't seen any, but maybe just haven't run across them. I personally dislike windows covered by the notch, inevitable there's some important information under it and I have to fiddle with the window to see it, so I'd prefer if there was a setting to never put anything under the notch.

**Comment by karpovpw** [comment]
> Are there other apps that do something like that?

I've seen some games that embrace the notch and full resolution of notched devices, like Minecraft and Life is Strange. It looks really cool with those thin display borders and the notch itself.

>I'd prefer if there was a setting to never put anything under the notch.

Well, that's the default behavior and an option to embrace the notch should be just that - an option.

The screen on 14" is comparably small, so adding 74px vertical pixels can actually increase readability a lot in some cases.

Apple documentation on auxiliaryTopLeftArea and auxiliaryTopRightArea looks simple, but I can't figure a way on how to draw content view using those properties.

**Comment by karpovpw** [comment]
@nickv2002 Hey man, any news? 

I just need some guidance with drawing content view and maybe I could help on this.

**Comment by nickv2002** [comment]
@karpovpw I've got a new M2 MacBook Air on its way to me soon. I'll test the current behavior around full-screen and notches and have a better idea of where to go. 

If you have a good idea of how it should work you don't need to wait for my guidance. Just make sure users have a way to control the behavior. Thanks for any/all help!

**Comment by nickv2002** [comment]
Got my new MacBook and did some tests: I think the current behavior where content does not flow around the notch is the correct default. I agree that it would be a nice feature in the prefs for full screen to flow content around (and behind) the notch.

My day job moved away from Mac dev work long ago so I'm far from good source on what APIs to use on this or even how to code it up. (I just keep a dev account open to sign/distribute fun projects like this.) 

However, on the design side I imagine a checkbox in the apps pref to swap between the default behavior and filling to the left and right of the notch would be sufficient and will be happy to review any PRs sent here that do that. Even a hidden plist/defaults write option would be fine initially.

**Comment by karpovpw** [comment]
Modern window management on macOS is hard for me to grasp, but the implemented custom solution in this project is a complete black box haha.

Still get to it from time to time though, any help appreciated.

**Comment by karpovpw** [comment]
Hey guys, I've just found this blog post: https://notes.alinpanaitiu.com/Fullscreen%20apps%20above%20the%20MacBook%20notch

Anyone willing to try it? I can't do it  :(

**Comment by nickv2002** [comment]
> Hey guys, I've just found this blog post: [notes.alinpanaitiu.com/Fullscreen%20apps%20above%20the%20MacBook%20notch](https://notes.alinpanaitiu.com/Fullscreen%20apps%20above%20the%20MacBook%20notch)
> 
> Anyone willing to try it? I can't do it :(

This is not the solution we're looking for: the suggestions on that blog post are about hacking existing apps and mucking around with sudo/SIP which is dangerous even if you know what you're doing. 

Instead there must be some tweaks we can make to the app and it's settings/entitlements in Xcode, but I don't know what those are or if App Store review would allow those.

**Comment by karpovpw** [comment]
> > Hey guys, I've just found this blog post: [notes.alinpanaitiu.com/Fullscreen%20apps%20above%20the%20MacBook%20notch](https://notes.alinpanaitiu.com/Fullscreen%20apps%20above%20the%20MacBook%20notch)
> > Anyone willing to try it? I can't do it :(
> 
> This is not the solution we're looking for: the suggestions on that blog post are about hacking existing apps and mucking around with sudo/SIP which is dangerous even if you know what you're doing.
> 
> Instead there must be some tweaks we can make to the app and it's settings/entitlements in Xcode, but I don't know what those are or if App Store review would allow those.

Well if you look closely, the article starts with a paragraph called “Old-style fullscreen” and these instructions are exactly what you are talking about. It’s just impossible for me to implement this custom solution for fullscreen rendering, so I was asking if anyone reading have the skills and is able to follow those instructions with provided code example how IINA does it. 

The hack you are mentioning is working ok and extra added pixels plus curved MacBook screen looks exceptional, especially on 14”.

**Comment by nickv2002** [comment]
Thanks for the more specific callout. Adding a [direct link to the IINA code](https://github.com/iina/iina/blob/fc66b27d50d0e98b056205867055f462e87828c9/iina/MainWindowController.swift#L1401-L1423) for future ref.

Not sure if this would get through app review though… none of the apps mentioned that have this are in the App Store AFAICT…

**Comment by DavidPhillipOster** [comment]
I'm tempted to take a crack at this. I don't see any problem with the [linked IINA code](https://github.com/iina/iina/blob/fc66b27d50d0e98b056205867055f462e87828c9/iina/MainWindowController.swift#L1401-L1423) passing app review.

My Mac mini has 3 displays, and I've often wanted full-screen mode to put something useful on all 3 displays. For Simple Comic, that would mean page 1 on the left portrait mode display, pages 2&3 on the large central display, and page 4 on the right portrait mode display. Where the "next page command" would display pages 5..8 and shift-next would display 2..5. No notch, though, so I can't test the part the [karpovpw](https://github.com/karpovpw) actually wants. For Simple Comic, there would be an additional checkbox in the Settings dialog to enable "custom fullscreen"

Before the turn of the century I wrote and published a photo-shoebox app, like Apple’s **Photos**, that used this technique for its slideshows.

**Comment by karpovpw** [comment]
I can test it on MacBook Pro 14”.
